product/specs/2022-01-views/03-modeling-view-columns.md

```diff
@@ -0,0 +1,28 @@
+---
+title: Views: Modeling View Columns
+description: 
+published: true
+date: 2022-01-13T19:49:54Z
+tags: 
+editor: markdown
+dateCreated: 2022-01-13T19:49:54Z
+---
+
+Here's how I think we should model view columns in our API and UI. Each heading represents an attribute of a View Column.
+
+### Data Type
+- **Definition**: This is the final data type of the content of the column after any computations etc. are applied.
+- **Allowed values**: same as Table data types.
+- **Required**. Data type should always be set, at the very least, we can treat unknown data types as text.
+
+### Data Sources
+ - **Definition**: This is the set of source columns that are used to generate the data in the current View column.
+- **Allowed values**: references to other Table or View columns, including other columns in the same View.
+- **Optional**: This could be empty for purely calculated columns (e.g. using the Postgres `random()` function and putting the output in a column)
+
+### Data Formula
```

Comment:
The formula would include the sources, yes.

I'm borrowing Element's UI to illustrate the idea, imagine Matrix channels are data sources.

Formula would be something like:
![Screen Shot 2022-01-20 at 3 52 16 PM](https://user-images.githubusercontent.com/287034/150420432-f1707138-4358-41f7-8a1d-f8826bca2706.png)

Sources would just be list of variables used in the formula:
![Screen Shot 2022-01-20 at 3 51 34 PM](https://user-images.githubusercontent.com/287034/150420492-c6233748-c84e-46f6-92a7-f287595b9b7f.png)

Reply:
I like how you made that illustration, Kriti. Creative. I'm leaning towards dropping Data Sources, and querying the Data Formula for what columns it references. That way there would be a single source of truth for this. Having Data Sources at the same level of interface as Data Formula is a bit in conflict with the fact that one is derived from the other.

Reply:
We won't always have a Formula, e.g. a view that's built on the query `SELECT Title, Release Year, Rating FROM Movies;` won't have any Formula for `Title`, `Release Year` or `Rating`. It will have a Source for each column, though. 

We also may not have a Source for a column with a Formula - e.g. a column might be generated using the `RANDOM()` function, which would be the Formula. There's no source data for it, though.